unittest/src/allow_db_test.rs

```diff
@@ -0,0 +1,104 @@
+//! Unit test cases for functionality in allow_db.
+
+use crate::allow_db::*;
+use core::cell::Cell;
+
+// Utility to call insert_ro_buffer with a slice.
+// Safety: insert_ro_slice does not prevent RoAllowBuffer from outliving slice.
+// Instead, the caller must make sure its use patterns don't cause invalid
```

Comment:
This seems...dangerous? Would heap allocation of RoAllowBuffer allow us to tie its lifetime to the passed in slice?

Reply:
Yes, definitely dangerous. The problem is the validity period for the slice is determined at runtime. E.g.:

```rust
let mut buffer: [u8; 8] = [0; 8];
let mut kernel = fake::Kernel::new();

// Creates a new RoAllowBuffer, which I'll call A.
FakeSyscalls::allow_ro_static(/*driver_num=*/ 1, /*buffer_num=*/ 2, &buffer);

// Does NOT make the previous RoAllowBuffer invalid, because buffer_num is different.
FakeSyscalls::allow_ro_static(/*driver_num=*/ 1, /*buffer_num=*/ 3, &[]);

// Makes RoAllowBuffer A invalid
FakeSyscalls::allow_ro_static(/*driver_num=*/ 1, /*buffer_num=*/ 2, &[]);
```

I don't think we have a way to enforce safety at compile time. Instead, the Allow implementations in `fake::Kernel` must be trusted to implement TRD 104 correctly.